### DISCUSSION TITLE
How do I solve BUFFER_BINDINGS_NOT_16_BYTE_ALIGNED issue?

### DISCUSSION THREAD

**Discussion by spencerkohan** [gfx-rs/wgpu#3116] [Q&A] [answered]
I'm trying to get a wgpu renderer to deploy on the web.

I am running into this error:

> Caused by:
>     In Device::create_render_pipeline
>       note: label = `RenderProcess`
>      Downlevel flags BUFFER_BINDINGS_NOT_16_BYTE_ALIGNED are required but not supported on the device.

I understand this has to do with buffer alignment, but which buffers exactly would this refer to, and how can I fix this issue?  I.e. I have a uniform buffer binding - would this refer to that, or else to my vertex buffer layouts?

**Comment by grovesNL** [comment, marked as answer]
The idea behind this error is that the WebGL backend requires uniform buffers to be 16-byte aligned. To support WebGL, you'll need to add padding somewhere in the uniform buffer (e.g., probably at the end). 

Your vertex buffer layouts should be fine.

A better error message is being worked on in #3099

**Reply by kpreid** [reply]
A clarification of something that bit me when I was dealing with this restriction: This error is given about the **uniform bindings in the shader**, not about the actual size of the buffer you attempt to bind. (Note that the error comes from creating the render pipeline, at which point the buffers aren't involved yet.) So, you have to modify your shader code, not your Rust code. Of course, the two must match, but if you're editing the wrong one and wondering why nothing ever changes…
